bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java

```diff
@@ -821,34 +841,37 @@ protected void release(boolean wait, boolean writer)
       final int refCount = this.refCount.decrementAndGet();
       if (canEvict(refCount, writer)) {
         assert (next != null);
-        final Runnable storer = getStorer(data, readingOffset, writingOffset, storage);
-        if (future != null && future.cancel(false)) {
-          logger.debug("Block {} future is cancelled", this);
-        }
-        final int numberOfInMemBlockPermits = DataList.this.numberOfInMemBlockPermits.get();
-        if (wait && numberOfInMemBlockPermits == 0) {
-          future = null;
-          storer.run();
-        } else if (numberOfInMemBlockPermits < MAX_COUNT_OF_INMEM_BLOCKS / 2) {
-          future = storageExecutor.submit(storer);
+        if (storage == null) {
+          final Runnable storer = getStorer(data, readingOffset, writingOffset, storage);
+          if (future != null && future.cancel(false)) {
+            logger.debug("Block {} future is cancelled", this);
+          }
+          final int numberOfInMemBlockPermits = DataList.this.numberOfInMemBlockPermits.get();
+          if (wait && numberOfInMemBlockPermits == 0) {
+            future = null;
+            storer.run();
+          } else if (numberOfInMemBlockPermits < MAX_COUNT_OF_INMEM_BLOCKS / 2) {
+            future = storageExecutor.submit(storer);
+          } else {
+            future = null;
+          }
         } else {
-          future = null;
+          // Resume checks number of in memory permits, always sending max when spooling is disabled
+          resumeSuspendedClients(MAX_COUNT_OF_INMEM_BLOCKS);
         }
       }
     }
 
     private boolean canEvict(final int refCount, boolean writer)
     {
-      if (refCount == 0 && storage != null) {
+      if (refCount == 0) {
         if (backPressureEnabled) {
           if (!writer) {
             boolean evict = true;
             int blocks = 0;
             // Search backwards from current block as opposed to searching forward from first block till current block as
-
-            // it is more likely to find the match quicker. No need to search more than maximum number of in memory
-            // permits.
-            for (Block temp = this.prev; (blocks < (MAX_COUNT_OF_INMEM_BLOCKS - 1)) && (temp != null); temp = temp.prev, ++blocks) {
+            // it is more likely to find the match quicker.
+            for (Block temp = this.prev; temp != null; temp = temp.prev, ++blocks) {
```

Comment:
variable `blocks` is not used. When storage != null, should it iterate only MAX_COUNT_OF_INMEM_BLOCK - 1?